File: solutions/ruby/sum-of-multiples/1/sum_of_multiples.rb
class SumOfMultiples

  MULTIPLE_OF = ->(factors, number) do
    factors.any? { |factor|
      factor.zero? ? false : number.modulo(factor).zero?
    }
  end

  private

  attr_accessor :factors

  def initialize(*factors)
    self.factors = factors
  end

  public

  def to(upper_bound)
    (0...upper_bound).sum do |number|
      MULTIPLE_OF.(factors, number) ? number : 0
    end
  end

end

File: solutions/elixir/sum-of-multiples/1/lib/sum_of_multiples.ex
defmodule SumOfMultiples do

  def to(limit, factors) when is_list(factors) do
    Enum.reduce(1..(limit - 1), 0, fn number, sum ->
      if multiple?(factors, number) do
        number + sum
      else
        sum
      end
    end)
  end

  defp multiple?(factors, number) do
    Enum.any?(factors, fn factor ->
      factor != 0 && rem(number, factor) == 0
    end)
  end

end
